File: rectangle_cutting/materials/partion.cpp
#include <bits/stdc++.h>
using namespace std;

typedef array<array<int, 41>, 41> matrix;
matrix ta;
int n, m, i, j;
int MIN, tp;

void block(matrix a, int h, int k, int c)
{
    int l;
    c = c + 1;
    if(c < MIN)
    {
        l = 0;
        while(h + l <= m && k + l <= n && a[h + l][k] == 0)
        {
            l = l + 1;
            for(i = 0; i < l; i++)
                for(j = 0; j < l; j++)
                    a[h + i][k + j] = l;
            if(h + l <= m && a[h + l][k] == 0)
                block(a, h + l, k, c);
        }
        if(h + l == m + 1 || a[h + l][k] != 0)
        {
            while(a[h][k] != 0 && k <= n)
            {
                h = h + a[h][k];
                if(h == m + 1)
                {
                    h = 1;
                    k = k + 1;
                }
            }
            if(k <= n)
                block(a, h, k, c);
            else
            {
                if(c < MIN)
                {
                    MIN = c;
                    ta = a;
                }
            }
        }
    }
}

int f(int a, int b)
{
    int q, r, sum;
    sum = 0;
    while(b != 0)
    {
        q = a / b;
        r = a % b;
        sum = sum + q;
        a = b;
        b = r;
    }
    return sum;
}

signed main()
{
    matrix a;
    cin >> n >> m;
    for(j = 1; j <= n; j++)
        for(i = 1; i <= m; i++)
            a[i][j] = 0;
    tp = f(n, m);
    cout << "F = " << tp << "\n";
    MIN = tp;
    block(a, 1, 1, 0);
    if(MIN == tp)
        cout << "NEVER BETTER\n";
    else
    {
        cout << "MIN = " << MIN << "\n";
        cout << "THE MATRIX IS\n";
        for(j = 1; j <= n; j++)
        {
            for(i = 1; i <= m; i++)
                cout << setw(3) << ta[i][j];
            cout << "\n";
        }
    }
}

File: rectangle_cutting/materials/partion.pas
program partion;

type matrix = array[1..20, 1..20] of integer;
var a, ta : matrix;
    m, n, i, j : integer;
    min, tp : integer;
procedure block(a : matrix; h, k, c : integer);
    var l : integer;
    
begin
    c := c + 1;
    if c < min then
        begin
            l := 0;
            while (h + l <= m) and (k + l <= n) and (a[h + l, k] = 0) do
                begin
                    l := l + 1;
                    for i := 0 to l - 1 do
                        for j := 0 to l - 1 do
                            a[h + i, k + j] := l;
            if (h + l <= m) and (a[h + l, k] = 0) then
                block(a, h + l, k, c)
        end;
    if (h + l = m + 1) or (a[h + l, k] <> 0) then
        begin
            while (a[h, k] <> 0) and (k <= n) do
                begin
                    h := h + a[h, k];
                    if h = m + 1 then
                        begin
                            h := 1;
                            k := k + 1;
                        end
                end;
            if k <= n then block(a, h, k, c)
            else
                begin
                    if c < min then
                        begin
                            min := c;
                            ta := a;
                        end
                end
        end
    end
end;
                
function f(a, b : integer) : integer;
var q, r, sum : integer;
begin
    sum := 0;
    while b <> 0 do
        begin
            q := a div b;
            r := a mod b;
            sum := sum + q;
            a := b;
            b := r;
        end;
    f := sum;
end;

begin {main}
    for j := 1 to n do
        for i := 1 to m do
            a[i, j] := 0;

readln(n); readln(m);
    tp := f(m, n);
    writeln('F = ', tp);
    min := tp;
    block(a, 1, 1, 0);
    if min = tp then
        writeln('NEVER BETTER')
    else    
    begin
        writeln('MIN = ', min);
        writeln('THE MATRIX IS');
        writeln;
        for j := 1 to n do
            begin
                for i := 1 to m do
                    write(ta[i, j] : 3);
                writeln
            end
    end
end.